Sources/Testing/Attachments/Test.Attachment.swift

```diff
@@ -0,0 +1,300 @@
+    ///     to a test report or to disk. If `nil`, the testing library attempts
+    ///     to derive a reasonable filename for the attached value.
+    ///   - sourceLocation: The source location of the attachment.
+    public init(
+      _ attachableValue: some Attachable & Sendable & Copyable,
+      named preferredName: String? = nil,
+      sourceLocation: SourceLocation = #_sourceLocation
+    ) {
+      self.attachableValue = attachableValue
+      self.preferredName = preferredName ?? Self.defaultPreferredName
+      self.sourceLocation = sourceLocation
+    }
+
+    /// A filename to use when writing this attachment to a test report or to a
+    /// file on disk.
+    ///
+    /// The value of this property is used as a hint to the testing library. The
+    /// testing library may substitute a different filename as needed. If the
+    /// value of this property has not been explicitly set, the testing library
+    /// will attempt to generate its own value.
+    public var preferredName: String
+  }
+}
+
+// MARK: -
+
+extension Test.Attachment {
+  /// Attach this instance to the current test.
+  ///
+  /// An attachment can only be attached once.
+  public consuming func attach() {
+    Event.post(.valueAttached(self))
+  }
+}
+
+// MARK: - Non-sendable and move-only attachments
+
+/// A type that stands in for an attachable type that is not also sendable.
+private struct _AttachableProxy: Test.Attachable, Sendable {
+  /// The result of `withUnsafeBufferPointer(for:_:)` from the original
+  /// attachable value.
+  var encodedValue = [UInt8]()
+
+  var estimatedAttachmentByteCount: Int?
+
+  func withUnsafeBufferPointer<R>(for attachment: borrowing Test.Attachment, _ body: (UnsafeRawBufferPointer) throws -> R) throws -> R {
+    try encodedValue.withUnsafeBufferPointer(for: attachment, body)
+  }
+}
+
+extension Test.Attachment {
+  /// Initialize an instance of this type that encloses the given attachable
+  /// value.
+  ///
+  /// - Parameters:
+  ///   - attachableValue: The value that will be attached to the output of
+  ///     the test run.
+  ///   - preferredName: The preferred name of the attachment when writing it
+  ///     to a test report or to disk. If `nil`, the testing library attempts
+  ///     to derive a reasonable filename for the attached value.
+  ///   - sourceLocation: The source location of the attachment.
+  ///
+  /// When attaching a value of a type that does not conform to both `Sendable`
+  /// and `Copyable`, the testing library encodes it as data immediately. If the
+  /// value cannot be encoded and an error is thrown, that error is recorded as
+  /// an issue in the current test and the resulting instance of
+  /// ``Test/Attachment`` is empty.
+  @_disfavoredOverload
+  public init(
+    _ attachableValue: borrowing some Test.Attachable & ~Copyable,
+    named preferredName: String? = nil,
+    sourceLocation: SourceLocation = #_sourceLocation
```

Comment:
This comment isn't related to this overload, but it prompted the idea…

Can we offer a way to attach a `Codable` value which doesn't itself conform to `Test.Attachable`? Perhaps via another overload of `Test.Attachment.init` which accepts `some Encodable`? That might require a different `_AttachableProxy`-esque helper type to wrap the codable value and conform to `Attachable`

Reply:
Patience, grasshopper. (Yes, that's coming in the next PR after this one. We need Foundation to make it work.)